biz.aQute.bnd.reporter/src/biz/aQute/bnd/reporter/plugins/entries/bundle/GogoPlugin.java

```diff
@@ -0,0 +1,375 @@
+					if (optional.isPresent()) {
+						gogoScopeDTO = optional.get();
+
+					} else {
+						gogoScopeDTO = new GogoScopeDTO();
+						gogoScopeDTO.title = scope;
+						gogoScopeDTO.functions = new ArrayList<GogoFunctionDTO>();
+
+					}
+
+					for (final String function : functions) {
+
+						Optional<GogoFunctionDTO> optionalF = gogoScopeDTO.functions.stream()
+							.filter(dto -> function.equals(dto.title))
+							.findFirst();
+
+						GogoFunctionDTO gogoFunctionDTO = null;
+
+						if (optionalF.isPresent()) {
+							gogoFunctionDTO = optionalF.get();
+						} else {
+							gogoFunctionDTO = new GogoFunctionDTO();
+							gogoFunctionDTO.title = function;
+							gogoFunctionDTO.methods = new ArrayList<GogoMethodDTO>();
+						}
+						String pathClazz = implementationClass.replace(".", "/") + ".class";
+						Resource resClazz = jar.getResource(pathClazz);
+						Analyzer analyzer = new Analyzer();
+						Clazz clazz = new Clazz(analyzer, null, resClazz);
+						clazz.parseClassFile(resClazz.openInputStream());
+						gogoFunctionDTO.methods.addAll(createMethodDTO(clazz, function));
+
+						gogoScopeDTO.functions.add(gogoFunctionDTO);
+					}
+					gogos.add(gogoScopeDTO);
+				}
+			} else {
+				if (!path.contains("*")) {
+					_reporter.warning("xml component file not found at path %s", path);
+				}
+			}
+		}
+		return !gogos.isEmpty() ? gogos : null;
+	}
+
+	private Collection<GogoMethodDTO> createMethodDTO(final Clazz clazz, final String function) throws Exception {
+
+		clazz.parseClassFile();
+
+		return clazz.methods()
+			.filter(m -> function.equals(m.getName()))
+			.map(m -> {
+
+				GogoMethodDTO gogoMethodDTO = new GogoMethodDTO();
+				gogoMethodDTO.title = m.getName();
+				gogoMethodDTO.parameters = Lists.newArrayList();
+
+				Optional<Annotation> oDescriptorAnnotation = m.annotations("*")
+					.filter(a -> ANNOTOATION_NAME_DESCRIPTOR.equals(a.getName()
+						.toString()))
+					.findFirst();
+
+				if (oDescriptorAnnotation.isPresent()) {
+
+					gogoMethodDTO.description = oDescriptorAnnotation.get()
+						.get("value");
+				}
+
+				Descriptor d = m.getDescriptor();
+				TypeRef[] trs = d.getPrototype();
+
+				MethodParameter[] mps = m.getParameters();
+
+				List<ParameterAnnotation> methodParameters = m.parameterAnnotations("*")
+					.collect(Collectors.toList());
+
+				for (int i = 0; i < trs.length; i++) {
+
+					GogoParameterDTO gogoParameterDTO = new GogoParameterDTO();
+					gogoParameterDTO.order = i;
+
+					if (mps != null && i < mps.length) {
+						gogoParameterDTO.title = mps[i].getName();
+					}
+
+					for (ParameterAnnotation parameterAnnotation : methodParameters) {
+
+						if (parameterAnnotation.parameter() == i
+							&& ANNOTOATION_NAME_DESCRIPTOR.equals(parameterAnnotation.getName()
+								.toString())) {
+
+							gogoParameterDTO.description = parameterAnnotation.get("value");
+						}
+
+						if (parameterAnnotation.parameter() == i
```

Comment:
Isn't this the same `if` test as immediately above?

Reply:
done, thank you.

Reply:
done